### PR TITLE
Optimize pipelines

### PR DESCRIPTION
**Pull request by anthonydresser**
Adds a few different optimization to reduce our release pipeline duration.

- Add a rebuilt linux container to avoid installing linux dependencies on every job on linux

- Precompile all our code once then reuse that in packaging for each os

- Use azure hosted windows machines for building windows packages then add a testing job that takes that drop and runs the test on our hosted machines.

- Adds server packages to our drops

Its hard to know what the actual time saved would be, but something like ~1 hour 20 minute pipeline runs to ~40 minute pipeline runs.

### CODE REVIEW

File: scripts/test-extensions-unit.sh [hunk 1]
```diff
@@ -21,49 +22,62 @@ if [[ "$ADS_TEST_GREP" == "" ]]; then
...
+
 cd $ROOT
 echo $VSCODEUSERDATADIR
 echo $VSCODEEXTDIR
 
 echo ***************************************************
 echo *** starting admin tool extension windows tests ***
 echo ***************************************************
-./scripts/code.sh --nogpu --extensionDevelopmentPath=$ROOT/extensions/admin-tool-ext-win --extensionTestsPath=$ROOT/extensions/admin-tool-ext-win/out/test --user-data-dir=$VSCODEUSERDATADIR --extensions-dir=$VSCODEEXTDIR
+"$INTEGRATION_TEST_ELECTRON_PATH" $LINUX_NO_SANDBOX --nogpu --extensionDevelopmentPath=$ROOT/extensions/admin-tool-ext-win --extensionTestsPath=$ROOT/extensions/admin-tool-ext-win/out/test --user-data-dir=$VSCODEUSERDATADIR --extensions-dir=$VSCODEEXTDIR
```
Comment (Charles-Gagnon) [inline scripts/test-extensions-unit.sh]: Not doing this for the batch file version as well?

File: build/azure-pipelines/linux/sql-product-build-linux.yml [hunk 2]
```diff
@@ -156,6 +175,20 @@ steps:
       SourceFolder: '$(Build.SourcesDirectory)/../vsix'
       TargetFolder: '$(Build.ArtifactStagingDirectory)/vsix'
 
+  - script: |
+      set -e
+      docker build -t azuredatastudio-server -f build/azure-pipelines/docker/Dockerfile $(agent.builddirectory)/azuredatastudio-reh-linux-x64
+      docker save azuredatastudio-server | gzip > azuredatastudio-server-docker.tar.gz
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/linux/sql-product-build-linux.yml]: What's this for?

File: scripts/test-extensions-unit.sh (hunk 1 above)
Comment (anthonydresser) [reply scripts/test-extensions-unit.sh]: It wasn't necessary for this change, but I can do that.

File: build/azure-pipelines/linux/sql-product-build-linux.yml (hunk 2 above)
Comment (anthonydresser) [reply build/azure-pipelines/linux/sql-product-build-linux.yml]: remote server feature.

File: extensions/integration-tests/src/objectExplorer.test.ts [hunk 3]
```diff
@@ -87,7 +87,7 @@ class ObjectExplorerTester {
 		let expectedActions: string[];
 		// Properties comes from the admin-tool-ext-win extension which is for Windows only, so the item won't show up on non-Win32 platforms
 		if (process.platform === 'win32') {
-			expectedActions = ['Manage', 'New Query', 'New Notebook', 'Disconnect', 'Delete Connection', 'Refresh', 'Data-tier Application wizard', 'Launch Profiler', 'Properties'];
```
Comment (alanrenmsft) [inline extensions/integration-tests/src/objectExplorer.test.ts]: why are we removing these 3 items here?

Comment (anthonydresser) [reply extensions/integration-tests/src/objectExplorer.test.ts]: Our shipped product doesn't have those extensions, so this test was failing.

Comment (alanrenmsft) [reply extensions/integration-tests/src/objectExplorer.test.ts]: please make the else branch consistent with windows for now.

**Pull request merged**